### ISSUE TITLE
INTERNAL ERROR: Failed to apply GPUdelegate to custom tflite model

### ISSUE THREAD

**Issue by prat96** [tensorflow/tensorflow#30303]
**System information**
- Have I written custom code: Yes, using a custom tflite model and Android app to run on my devices.
- OS Platform and Distribution: Ubuntu 18.04
- Mobile device if the issue happens on mobile device: Asus zenphone M2, Snapdragon 820, 605 dev-kits.
- TensorFlow installed from: nightly AAR (org.tensorflow:tensorflow-lite:0.0.0-nightly)
- TensorFlow-gpu installed from:  nightly AAR (org.tensorflow:tensorflow-lite-gpu:0.0.0-nightly)
- Python version: 3.6.7
- CUDA/cuDNN version: 10.0
- GPU model and memory: GTX 1060 6GB

**Describe the current behavior**
I've written a simple CNN to count number of fingers, works perfectly on my laptop (both the .h5 and tflite version). When I load the .tflite model on Android it fails saying;

` Caused by: java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: TfLiteGpuDelegate Prepare: Shader compilation failed: ERROR: 0:6: 'unknown' : not a legal layout qualifier id`

What exactly does this error mean? 

**Describe the expected behavior**
GPUdelegate should load correctly and run inference on my Android devices.

**Code to reproduce the issue**
My custom app is based on the TensorFlow Lite Android image classification example. The model loads and runs fine on CPU. The only difference in the code is the following lines:

```
tfliteOptions = new Interpreter.Options();
gpuDelegate = new GpuDelegate();
tfliteOptions.addDelegate(gpuDelegate);
```

**Other info / logs**
GPUdelegate works fine if I try to run models like MobilenetV2, without any change in code. 
After much googling, [found this on StackOverflow](https://stackoverflow.com/questions/55794791/tensorflow-lite-gpu-delegate-failure)

So, is it an issue with the output layers in my model? I'll include the model graph. 
Please help!

**Traceback:**
```
2019-07-02 10:19:12.153 15591-15613/com.prat96.tflite I/Adreno: ERROR: 0:6: 'unknown' : not a legal layout qualifier id 
    ERROR: 0:6: 'unknown' : Syntax error:  syntax error
    INTERNAL ERROR: no main() function!
    ERROR: 2 compilation errors.  No code generated.
2019-07-02 10:19:12.156 15591-15613/com.prat96.tflite E/AndroidRuntime: FATAL EXCEPTION: pool-1-thread-1
    Process: com.prat96.tflite, PID: 15591
    java.lang.RuntimeException: Error initializing TensorFlow!
        at com.prat96.tflite.MainActivity$3.run(MainActivity.java:133)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: TfLiteGpuDelegate Prepare: Shader compilation failed: ERROR: 0:6: 'unknown' : not a legal layout qualifier id 
    ERROR: 0:6: 'unknown' : Syntax error:  syntax error
    INTERNAL ERROR: no main() function!
    ERROR: 2 compilation errors.  No code generated.
    
    Node number 11 (TfLiteGpuDelegate) failed to prepare.
    
        at org.tensorflow.lite.NativeInterpreterWrapper.applyDelegate(Native Method)
        at org.tensorflow.lite.NativeInterpreterWrapper.init(NativeInterpreterWrapper.java:83)
        at org.tensorflow.lite.NativeInterpreterWrapper.<init>(NativeInterpreterWrapper.java:60)
        at org.tensorflow.lite.Interpreter.<init>(Interpreter.java:224)
        at com.prat96.tflite.TensorFlowImageClassifier.create(TensorFlowImageClassifier.java:57)
        at com.prat96.tflite.MainActivity$3.run(MainActivity.java:126)
        	... 3 more
2019-07-02 10:19:12.168 15591-15613/? I/Process: Sending signal. PID: 15591 SIG: 9
```

**Comment by prat96** [comment]
![tf2_fingers](https://user-images.githubusercontent.com/19527450/60494011-aa7fe480-9cb6-11e9-885e-e8dad29c2daf.png)

**Comment by impjdi** [comment]
@prat96 

Is it possible to share the network?  You don't need to share the weights/bias.  I only need a TFLite file with the correct network architecture.

**Comment by prat96** [comment]
@impjdi Absolutely, I apologize for not putting that up in the first place!

[Link to .h5 file, .tflite file and some test images](https://drive.google.com/open?id=1eb5LMDsZW_sgUSjljpCLiX1z_bGYfBEW)

**Comment by impjdi** [comment]
@prat96 

Thanks.  Can you try to insert a reshape op right before the first FC, reshaping the tensor from [1, 6, 6, 32] to [1, 1152]?

**Comment by prat96** [comment]
Ok, I have done this by adding;

**Comment by prat96** [comment]
ok, I have done this by adding;
``` 
tf.keras.layers.Reshape((1, 1152)),
tf.keras.layers.Flatten(),
tf.keras.layers.Dense(128, activation='relu'),
....
```
Aaaanndd now it works perfectly! 
@impjdi Could you please tell me in detail why this step is nescessary? I want to understand so that I can develop other models, keeping this in mind. 
Thanks so much for your help!

**Comment by impjdi** [comment]
@prat96 

Ah, so the suspicion came from this:

[ B H W I ]  -> FC with weights [ O H W I ] and bias [ O ]  -> [ B H W O ]

I saw that your input and weights don't match, so I asked you to fix that.  On the CPU, [ 1 6 6 32 ] and [ 1 1 1 1152 ] have the same memory layout, so it may work just out of the box without a seg fault, but I think you got lucky that TFLite CPU op impl is not being strict.

Now for the GPU, it is also true that the memory layout for [ 1 6 6 32 ] and [ 1 1 1 1152 ] is also the same, but it looked like the shader ops were expecting something that matches.  And then it was not able to find that, and that's why they got "unknown" data type instead of, e.g. "rgba16f" in `std::string ToImageLayoutQualifier(DataType type)` in `object_accessor.cc`.

In the future, I would try to stick to canonical input/output dimensions.  Ideally, the error log should be more descriptive, so that you can take action, but it's a problem we have.  Philosophically, we designed the shader parts to be agnostic of neural net: they are just a pipeline of executing GPU shader programs.  At that point, all the information of tensor names etc. got lost, and we were not able to pinpoint (in the error log), which tensor was bad =/  Maybe we have to revisit that decision, or find other ways to embed extra information for the error log...

**Comment by impjdi** [comment]
@rmothukuru 

I was thinking of closing this.  Do you have anything in mind as you have re-assigned this to yourself?

**Comment by rmothukuru** [comment]
@impjdi ,
Thank you for confirming it with me. I have assigned it on my name to label the issue.
@prat96 ,
Can you please confirm if your issue is resolved and if we can close this issue.

**Comment by prat96** [comment]
@impjdi Aah, thats really interesting and makes a lot of sense...
Do you think it would help if I had enabled some padding options in the first input layers? 

Again, thanks a lot! :)

@rmothukuru Yes my issue is resolved for now.

**Comment by rmothukuru** [comment]
Are you satisfied with the resolution of your issue?<br> [Yes](https://docs.google.com/forms/d/e/1FAIpQLSfaP12TRhd9xSxjXZjcZFNXPGk4kc1-qMdv3gc6bEP90vY1ew/viewform?entry.85265664=Yes&entry.2137816233=30303)<br> [No](https://docs.google.com/forms/d/e/1FAIpQLSfaP12TRhd9xSxjXZjcZFNXPGk4kc1-qMdv3gc6bEP90vY1ew/viewform?entry.85265664=No&entry.2137816233=30303)
